### PR TITLE
[FLINK-8362][elasticsearch] shade all dependencies

### PR DESCRIPTION
**Pull request by NicoK**
## What is the purpose of the change

The Elasticsearch connectors have some dependencies that need to be available and should not conflict with other user or system code. Similarly to the cassandra connector and the S3 file systems, it should thus shade all its dependencies and become self-contained.

## Brief change log

- shade dependencies of `flink-connector-elasticsearch`, `flink-connector-elasticsearch2`, and `flink-connector-elasticsearch5` and relocate to our namespace
- add a test into `travis_mvn_watchdog.sh` to verify classes are shaded (one gap though: since `flink-connector-elasticsearch` and `flink-connector-elasticsearch-base` use the same package, we cannot really test that the latter is shaded correctly there)
- update documentation

## Verifying this change

This can be verified as follows:

In the shaded jar files,
- check for non-shaded `service` files (also their contents)
- check for non-shaded `.class` files (this test is automated in `travis_mvn_watchdog.sh`, but `flink-connector-elasticsearch` should be verified manually as well - see the note above)
- check for unnecessary files in `META-INF/maven` (only `org.apache.flink/flink-connector-elasticsearch<variant>_<scala_version>/` should remain)
- manually test the connectors and verify they are working without adding other dependencies to a user jar (TODO! @tzulitai since you have been working with the connector in the past, can you chime in here?)

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **no**
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive): **no**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: **no**
  - The S3 file system connector: **no**

## Documentation

  - Does this pull request introduce a new feature? **no**
  - If yes, how is the feature documented? **docs**

### CODE REVIEW

**Comment by StephanEwen** [comment]
I vaguely recall that there was a reason why we did not shade the Elasticsearch connectors initially. Maybe a license issue. @zentol and @tzulitai can you chime in here?

If we want to go ahead with this, few comments:
  - Not sure we should relocate the Flink elasticsearch code. Why is that needed?
  - Would be good to add for each dependency that is shaded should have a comment about the license. Some licenses are not compatible with shading.
  - Some licenses may need an additional license file or a mention in an extra bundled NOTICE file, see the `flink-s3-fs-presto`
  - In general, we cannot suppress all forwarded files from `META-INF` for dependencies, especiallylicense related files.

File: docs/dev/connectors/elasticsearch.md [hunk 1]
```diff
@@ -440,36 +440,7 @@ For the execution of your Flink program, it is recommended to build a
...
-                <transformers>
-                    <transformer implementation="org.apache.maven.plugins.shade.resource.ServicesResourceTransformer"/>
-                </transformers>
-            </configuration>
-        </execution>
-    </executions>
-</plugin>
-~~~
+Alternatively, you can put the connector's jar file into Flink's `lib/` folder to make it available
+system-wide, i.e. for all job being run.
```
Comment (tzulitai) [inline docs/dev/connectors/elasticsearch.md]: Not sure whether we should really recommend / mention this, as creating an uber jar is the usual recommended approach.

File: flink-connectors/flink-connector-elasticsearch/pom.xml [hunk 2]
```diff
@@ -93,6 +93,106 @@ under the License.
...
+								<relocation>
+									<pattern>com.fasterxml.jackson</pattern>
+									<shadedPattern>org.apache.flink.streaming.connectors.elasticsearch.shaded.org.elasticsearch.common.jackson</shadedPattern>
+								</relocation>
+								<relocation>
+									<pattern>com.spatial4j</pattern>
+									<shadedPattern>org.apache.flink.streaming.connectors.elasticsearch.shaded.com.spatial4j</shadedPattern>
+								</relocation>
+
+								<!-- relocate everything from the flink-connector-elasticsearch base project -->
```
Comment (tzulitai) [inline flink-connectors/flink-connector-elasticsearch/pom.xml]: Not sure if this is necessary. @NicoK can you comment on this?

**Comment by tzulitai** [comment]
I can't recall any license issues with shading the ES connectors ... will keep an extra eye and double check on that before merging this.

@NicoK thanks a lot for your work on this. I'll first proceed to manually test this, and after that get back to the licensing issues that Stephan mentioned.

**Pull request closed without merging**